packages/libs/error-stack/src/fmt/hook.rs

```diff
@@ -504,18 +504,37 @@ impl Hooks {
         &mut self,
         hook: impl for<'a> Fn(&T, &mut HookContext<'a, T>) -> Vec<Emit> + Send + Sync + 'static,
     ) {
-        let inner = self.inner.get_or_insert_with(BTreeMap::new);
-
-        inner.insert(
-            TypeId::of::<T>(),
-            Box::new(move |frame, ctx| {
-                // `.unwrap()` never fails here, because `Hooks` guarantees the function will never
-                // be called on an object which cannot be downcast.
-                let frame = frame.downcast_ref::<T>().unwrap();
-
-                hook(frame, ctx.cast())
-            }),
-        );
+        let type_id = TypeId::of::<T>();
+
+        let boxed_hook = Box::new(move |frame: &Frame, ctx: &mut HookContext<Frame>| {
+            #[cfg(nightly)]
+            {
+                frame
+                    .request_ref::<T>()
+                    .map(|val| hook(val, ctx.cast()))
+                    .or_else(|| {
+                        frame
+                            .request_value::<T>()
+                            .as_ref()
+                            .map(|val| hook(val, ctx.cast()))
+                    })
+                    .unwrap_or_default()
+            }
+
+            #[cfg(not(nightly))]
+            {
+                frame
+                    .downcast_ref::<T>()
+                    .map(|val| hook(val, ctx.cast()))
+                    .unwrap_or_default()
+            }
+        });
+
+        if let Some(hook) = self.inner.iter_mut().find(|(id, _)| *id == type_id) {
+            *hook = (type_id, boxed_hook);
+        } else {
+            self.inner.push((type_id, boxed_hook));
+        }
```

Comment:
This implies, that a hook, which replaces an old hook, is not appended to the vector but is replaced in place (see above)

Reply:
clarified in [d388a1b](https://github.com/hashintel/hash/pull/974/commits/d388a1b9f48f3acad5d61e3a7afd6dabf371897e)